### PR TITLE
feat(restart-inbound): give the possibility to the developer to plan a restart of the connector in case of failure

### PR DESCRIPTION
**Pull request by mathias-vandaele**
## Description

Proposal of solution for the restart of inbound connector, not tested yet

## Related issues

closes https://github.com/camunda/connectors/issues/3319

## Checklist

- [ ] PR has a **milestone** or the `no milestone` label.

### CODE REVIEW

**Comment by sbuettner** [comment]
Hey @mathias-vandaele, the ticket contains an implementation suggestion: https://github.com/camunda/connectors/issues/3319 to re-use the existing cancellation mechanism. Would be interesting to understand you see the need to introduce another method?

**Comment by mathias-vandaele** [comment]
> Hey @mathias-vandaele, the ticket contains an implementation suggestion: #3319 to re-use the existing cancellation mechanism. Would be interesting to understand why you see the need to introduce another method?

@sbuettner everything still happens during the cancellation mechanism 

 ```
 @Override
  public void cancel(Throwable exception) {
    try {
      cancellationCallback.accept(exception);
    } catch (Throwable e) {
      LOG.error("Failed to deliver the cancellation signal to the runtime", e);
    }
    if (Objects.requireNonNull(exception) instanceof RestartException restartException) {
      reactivationCallback.accept(restartException.getBackoffTime());
    }
  }
```

A callback is declared in order to be executed after cancelation 👍

File: connector-runtime/connector-runtime-core/src/main/java/io/camunda/connector/runtime/core/error/RestartException.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.camunda.connector.runtime.core.error;
+
+import java.time.Duration;
+
+public class RestartException extends RuntimeException {
```
Comment (johnBgood) [inline connector-runtime/connector-runtime-core/src/main/java/io/camunda/connector/runtime/core/error/RestartException.java:21]: Does it make sense to reuse the `ConnectorRetryException` with the `retries` and `backoff` parameters?

**Comment by sbuettner** [comment]
@mathias-vandaele Ah, I was confused as I saw: ```reactivationCallback```.  Would it be simpler to re-use the existing cancellation callback and handle it in the consumer?

File: connector-runtime/connector-runtime-core/src/main/java/io/camunda/connector/runtime/core/error/RestartException.java (hunk 1 above)
Comment (sbuettner) [reply connector-runtime/connector-runtime-core/src/main/java/io/camunda/connector/runtime/core/error/RestartException.java:21]: @johnBgood Good idea although we would need a way to restart without defining the number of retries. We could add it. WDYT?

File: connector-runtime/connector-runtime-spring/src/main/java/io/camunda/connector/runtime/inbound/executable/InboundExecutableRegistryImpl.java [hunk 2]
```diff
@@ -28,37 +28,89 @@
...
 import org.slf4j.LoggerFactory;
 import org.springframework.scheduling.annotation.Scheduled;
 
 public class InboundExecutableRegistryImpl implements InboundExecutableRegistry {
 
+  private static final Logger LOG = LoggerFactory.getLogger(InboundExecutableRegistryImpl.class);
+  final Map<UUID, RegisteredExecutable> executables = new HashMap<>();
   private final BlockingQueue<InboundExecutableEvent> eventQueue;
+  private final ScheduledExecutorService reactivationScheduler =
+      Executors.newSingleThreadScheduledExecutor();
```
Comment (johnBgood) [inline connector-runtime/connector-runtime-spring/src/main/java/io/camunda/connector/runtime/inbound/executable/InboundExecutableRegistryImpl.java:52]: any reason to use a single thread executor here?

File: connector-runtime/connector-runtime-core/src/main/java/io/camunda/connector/runtime/core/error/RestartException.java (hunk 1 above)
Comment (johnBgood) [reply connector-runtime/connector-runtime-core/src/main/java/io/camunda/connector/runtime/core/error/RestartException.java:21]: Yep, either that or we add the retries count to the context and when throwing the exception later we use this number, similar to how it's done for Outbound connectors.

**Pull request closed without merging**